lib/Dialect/Arith/Conversions/ArithToModArith/ArithToModArith.cpp

```diff
@@ -108,6 +75,24 @@ struct ConvertConstant : public OpConversionPattern<mlir::arith::ConstantOp> {
   }
 };
 
+struct ConvertExt : public OpConversionPattern<mlir::arith::ExtSIOp> {
+  ConvertExt(mlir::MLIRContext *context)
+      : OpConversionPattern<mlir::arith::ExtSIOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      ::mlir::arith::ExtSIOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto result = b.create<mod_arith::ModSwitchOp>(
+        op.getLoc(), convertArithType(op.getType()), op.getIn());
```

Comment:
I see the problem: you should be using `adaptor.getIn()` instead of `op.getIn()`.

The `OpAdaptor` has the type-converted operands and result types stored on it, but when you pass the original types from `ExtSIOp`, it leaves them as integers, and the op fails to verify and it adds that unrealized conversion cast op to try to compensate.

Reply:
Damn; that's the difference between adaptor and op. Thanks very much for the help!